### ISSUE TITLE
Proposal: back button closes modal dialog

### ISSUE THREAD

**Issue by nolanlawson** [nolanlawson/pinafore#60]
This seems intuitive to me (especially on Android browsers), but have to check to make sure it meets a11y guidelines.

**Comment by algernon** [comment]
Please do this. Every other application on my Android phone closes modals when I press back, except for Pinafore, and this is mighty annoying.

The material design says this about [dismissing dialogs](https://material.io/design/components/dialogs.html#behavior):

> Dialogs may be dismissed by:
>  * Tapping outside of the dialog
>  * Tapping the “Cancel” button
>  * Tapping the system back button (Android only)

**Comment by nolanlawson** [comment]
I can't find any documents on web a11y (e.g. [WAI-ARIA practices on dialogs](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/alertdialog.html)) that recommends having them close with the back button. The Material Design document mentioned above may refer only to Android native apps; it's not clear from the description.

It seems to me that this is a convention in Android native apps but not iOS native apps or web apps. Since Pinafore is a web app, arguably it should follow web conventions even when running on Android. I agree it's counter-intuitive in many ways for Android users though.

I also tried to implement this (#466) and found that it was incredibly hairy because I basically needed to maintain a hashchange system where every dialog had a unique hash and could open/close itself on back/forward. E.g. you open a dialog, you press back, you press forward, you press back, etc. Also I had to make sure that Esc/X button/ clicking outside of the modal also navigates "back" which besides being awkward is also a bit weird conceptually. (Should the modal really be in the browser history?)

This feature is interesting and definitely has merits, but for the moment I'm going to say "won't fix." I wish there were a way to just say "when an Android user taps the back button, dismiss the dialog" but it doesn't seem that simple in the web context.

**Comment by nolanlawson** [comment]
Actually I take it back, it looks like the Mastodon web UI already implements this, without updating the hash. Pinafore should probably stick with Mastodon conventions.

**Comment by nolanlawson** [comment]
Bug filed on Sapper; I should be able to do this with a proper hashchange-driven modal, but the Sapper bug needs fixing first.
